dev/com.ibm.ws.concurrent.mp.cdi/src/com/ibm/ws/concurrent/mp/cdi/ConcurrencyCDIExtension.java

```diff
@@ -36,13 +56,129 @@
+    private final Map<String, Map<String, ManagedExecutorConfig>> beanMap = new HashMap<>(4);
+    private final Set<Throwable> deploymentErrors = new LinkedHashSet<>();
+
+    // Cannot use proper DS @Reference here because CDI Extensions are created/used as unmanaged objects
+    private CDIService cdiSvc;
+
+    public void processInjectionPoint(@Observes ProcessInjectionPoint<?, ManagedExecutor> event, BeanManager bm) {
+        Annotated injectionPoint = event.getInjectionPoint().getAnnotated();
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+            Tr.debug(tc, "processInjectionPoint " + injectionPoint);
+
+        // Skip if no @ManagedExecutorConfig or module is not CDI enabled
+        if (!injectionPoint.isAnnotationPresent(ManagedExecutorConfig.class))
+            return;
+        CDIService cdiSvc = getCDIService();
+        if (!cdiSvc.isCurrentModuleCDIEnabled()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                Tr.debug(tc, "Current module is not CDI enabled. Skipping further processing.");
+            return;
+        }
+
+        // Store configs in 3 tier map:  appName -> instanceName -> config
+        String appName = cdiSvc.getCurrentApplicationContextID();
+        Map<String, ManagedExecutorConfig> appBeanMap = beanMap.computeIfAbsent(appName, (e) -> new HashMap<>(8));
+        ManagedExecutorConfig config = injectionPoint.getAnnotation(ManagedExecutorConfig.class);
+
+        // Instance name is either @NamedInstance.value() or generated from fully-qualified class+field name
+        NamedInstance nameAnno = injectionPoint.getAnnotation(NamedInstance.class);
+        Member member = event.getInjectionPoint().getMember();
+        String name = nameAnno == null ? member.getDeclaringClass().getTypeName() + "." + member.getName() : nameAnno.value();
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-            Tr.debug(tc, "Registering MP Concurrency producers");
-        AnnotatedType<ManagedExecutorProducer> execType = bm.createAnnotatedType(ManagedExecutorProducer.class);
-        event.addAnnotatedType(execType, ManagedExecutorProducer.class.getName());
+            Tr.debug(tc, "InjectionPoint " + name + " in app " + appName + " has config " + config);
+        if (nameAnno == null)
+            event.configureInjectionPoint().addQualifiers(createNamedInstance(name));
+        ManagedExecutorConfig prevoiusConfig = appBeanMap.putIfAbsent(name, config);
+
+        // If 2 or more InjectionPoints define @NamedInstance("X") @ManagedExecutorConfig it is an error
+        if (prevoiusConfig != null) {
+            String msg = "ERROR: Found existing bean with name=" + name; // TODO NLS message
+            Tr.error(tc, msg);
+            deploymentErrors.add(new Throwable(msg));
+        }
+    }
+
+    private CDIService getCDIService() {
```

Comment:
Given that this class is a declarative services component, why isn't CDIService bound as a dependency?